### ISSUE TITLE
"Cold T" error with ne120 on pm-gpu

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/scream#2061]
Using scream master of Nov 23rd (after last upstream merge), and with `--res
 ne120pg2_r0125_oRRS18to6v3 --mach pm-gpu --compiler gnugpu --compset F2010-SCREAMv1`, I now see a "cold T" error.  The ne120 attempts still use default 72 vertical levels. I did NOT see this error at ne120 with repos of Oct29th and older -- we know now it was because of my using `Ck=1.0`.  This case failed after step 166 with:

```
30: terminate called after throwing an instance of 'std::logic_error'
30:   what():  /global/cfs/cdirs/e3sm/ndk/repos/se45-nov23/components/eamxx/src/share/atm_process/atmosphere_process.cpp:256: FAIL:
30: false
30: Error! Failed post-condition property check (cannot be repaired).
30:   - Atmosphere process name: Dynamics
30:   - Property check name: T_mid within interval [130, 500]
30:   - Atmosphere process MPI Rank: 30
30:   - Message: Check failed.
30:   - check name: T_mid within interval [130, 500]
30:   - field id: T_mid[Physics PG2] <double:COL,LEV>(3600,72) [K]
30:   - minimum:
30:     - value: 121.83
30:     - entry: (106868,71)
30:     - lat/lon: (31.4413, 102.563)
30:   - maximum:
30:     - value: 302.883
30:     - entry: (93902,71)
30:     - lat/lon: (11.2015, 101.813)

/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/se45-nov23/f120.F2010-SCREAMv1.ne120pg2_r0125_oRRS18to6v3.se45-nov23.gnugpu.5d.n024a4xX18888.wr.s16.gpull
```

The only source change I had to this repo was Ck=1.0:

```
--- a/components/scream/src/physics/shoc/impl/shoc_compute_shr_prod_impl.hpp
+++ b/components/scream/src/physics/shoc/impl/shoc_compute_shr_prod_impl.hpp
@@ -24,7 +24,8 @@ void Functions<S,D>
   const uview_1d<Spack>&       sterm)
 {
   // Turbulent coefficient
-  const Scalar Ck_sh = 0.1;
+  //ndk const Scalar Ck_sh = 0.1;
+  const Scalar Ck_sh = 1.0;
 
```

This issue https://github.com/E3SM-Project/E3SM/issues/6817 shows "cold T" error with ne256. But that was "resolved" (we hope) with Ck=1.0.

**Comment by ndkeen** [comment]
I can repeat the error with other node counts at ne120.  And at least one case on pm-cpu has same issue.

**Comment by whannah1** [comment]
@ndkeen did you write any detailed output for this case? If not, would it be possible to rerun one of the failing cases with 3D output every time step so we can dig into the conditions leading up to the cold-T?

**Comment by ndkeen** [comment]
I ran 3 months of ne120 using default source in master of Dec 5th.  That is, `Ck=0.1` as we now think changing this value is maybe not helping. I ran with similar outputs as our production ne1024 case. @crterai verified that we are writing the min T_mid values correctly.
`/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/se47-dec5/f120.F2010-SCREAMv1.ne120pg2_r0125_oRRS18to6v3.se47-dec5.gnugpu.3m.n032a4xX.ybo`

**Comment by ndkeen** [comment]
I was trying to run ne120 longer and hit this error after 5.5 months:

```
 96: WARNING: Tl1_1 has 1 values <= 0.  Resetting to minimum value.
 89: Bad dphi, dp3d, or vtheta_dp; label: 'CaarFunctorImpl::run TagPreExchange'; see hommexx.errlog.128.89
 89: Exiting...
 89: MPICH ERROR [Rank 89] [job id 3932109.0] [Wed Dec  7 16:24:50 2022] [nid003337] - Abort(102) (rank 89 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 102) - process 89
 89:
 89: aborting job:
 89: application called MPI_Abort(MPI_COMM_WORLD, 102) - process 89
 89: Kokkos::Cuda ERROR: Failed to call Kokkos::Cuda::finalize()
 89: Error: Trying to free an invalid pointer

/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/se47-dec5/f120.F2010-SCREAMv1.ne120pg2_r0125_oRRS18to6v3.se47-dec5.gnugpu.1y.n032a4xX.minToutput
```

**Comment by ndkeen** [comment]
@crterai made a plot of min T
![image](https://user-images.githubusercontent.com/5684727/207740322-71e1a7bd-5beb-490b-97d5-3b17e3ade0e0.png)

**Comment by ndkeen** [comment]
Using Jan4th repo, I ran this again with same number of nodes on pm-gpu.  Fails at same date, but the error mesg might be different.

```
 89: Warning: command line argument '--kokkos-num-devices' is deprecated. Use '--kokkos-map-device-id-by=mpi_rank' instead. Raised by Kokkos::initialize().
 89: terminate called after throwing an instance of 'std::logic_error'
 89:   what():  /global/cfs/cdirs/e3sm/ndk/repos/se49-jan4/components/eamxx/src/share/atm_process/atmosphere_process.cpp:313: FAIL:
 89: false
 89: Error! Failed post-condition property check (cannot be repaired).
 89:   - Atmosphere process name: Dynamics
 89:   - Property check name: NaN check for field T_mid
 89:   - Atmosphere process MPI Rank: 89
 89:   - Message: FieldNaNCheck failed.
 89:   - field id: T_mid[Physics PG2] <double:COL,LEV>(2700,72) [K]
 89:   - entry (264636,71)
 89:   - lat/lon: (-71.999225, 298.497298)
 89: 
 89:  *************************** INPUT FIELDS ******************************
 89: 
 89:   ------- INPUT FIELDS -------
 89:      T_mid<COL,LEV>(2700,72)
 89: 
 89:   T_mid(2567,:)
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, -nan, -nan, -nan, 
 89:     -nan, -nan, 
 89:  -----------------------------------------------------------------------
 89:      horiz_winds<COL,CMP,LEV>(2700,2,72)
 89: 
 89:   horiz_winds(2567,0,:)
 89:     182.694, 178.743, 164.6, 146.56, 131.493, 
 89:     113.449, 104.724, 95.7102, 88.7424, 80.2443, 
 89:     72.5874, 65.9015, 59.3869, 57.4037, 56.3147, 
 89:     54.3883, 51.5292, 47.9175, 44.5168, 42.6113, 
 89:     40.7554, 39.4643, 39.1098, 37.87, 36.1776, 
 89:     34.6962, 33.7756, 32.8687, 32.1771, 32.4635, 
 89:     34.1787, 35.7404, 36.1103, 35.4887, 35.5277, 
 89:     36.151, 37.2918, 38.0968, 38.3416, 38.3908, 
 89:     38.6769, 38.729, 38.2499, 36.3384, 33.1563, 
 89:     28.04, 22.2446, 16.9411, 12.8739, 10.8641, 
 89:     10.3676, 10.5586, 10.9046, 10.9018, 9.85478, 
 89:     8.40555, 6.58623, 4.85174, 3.10253, 1.38171, 
 89:     -0.253599, -1.73419, -2.9785, -3.81165, -4.5392, 
 89:     -4.29207, -4.20877, -4.11428, -3.79642, -3.50825, 
 89:     -3.41855, -3.22453, 
 89:   horiz_winds(2567,1,:)
 89:     41.0091, 34.7279, 38.371, 42.3312, 41.5417, 
 89:     42.7733, 39.1788, 37.1453, 37.5993, 36.5454, 
 89:     32.5339, 28.5897, 25.3212, 24.9514, 25.9568, 
 89:     26.6306, 27.114, 26.8786, 24.9836, 21.845, 
 89:     19.4803, 16.2378, 13.6197, 12.019, 9.9433, 
 89:     7.91011, 7.06119, 8.20501, 9.96926, 10.0893, 
 89:     9.97738, 9.58503, 7.64244, 4.90279, 2.80819, 
 89:     1.95955, 2.29342, 3.11755, 3.28172, 2.86035, 
 89:     1.98066, 1.64826, 1.56496, 1.37929, 1.40116, 
 89:     2.08861, 2.95577, 3.46225, 3.75012, 4.5506, 
 89:     5.46848, 5.98011, 6.13433, 5.97825, 5.7605, 
 89:     5.6641, 5.67487, 5.79228, 5.97832, 6.23399, 
 89:     6.60159, 7.09906, 7.73102, 8.53105, 9.36315, 
 89:     9.52364, 9.24704, 8.84944, 8.09782, 7.29806, 
 89:     6.80308, 6.12072, 
 89:  -----------------------------------------------------------------------
 89:      o3_volume_mix_ratio<COL,LEV>(2700,72)
 89: 
 89:   o3_volume_mix_ratio(2567,:)
 89:     1.91483e-07, 2.82726e-07, 4.17448e-07, 6.16365e-07, 9.10068e-07, 
 89:     1.34372e-06, 1.98402e-06, 2.92942e-06, 4.32531e-06, 5.48176e-06, 
 89:     5.48284e-06, 5.34783e-06, 4.92234e-06, 4.48612e-06, 4.16686e-06, 
 89:     3.83683e-06, 3.59221e-06, 3.31163e-06, 2.94947e-06, 2.57175e-06, 
 89:     2.2278e-06, 1.89477e-06, 1.65329e-06, 1.44071e-06, 1.25519e-06, 
 89:     1.08486e-06, 9.48895e-07, 8.21839e-07, 7.34003e-07, 6.48835e-07, 
 89:     5.87928e-07, 5.24116e-07, 4.61335e-07, 3.95203e-07, 3.26546e-07, 
...
/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/se49-jan4/f120.F2010-SCREAMv1.ne120pg2_r0125_oRRS18to6v3.se49-jan4.gnugpu.n032a4xX
```

**Comment by ndkeen** [comment]
Using @brhillman branch `brhillman/add-ne120pg2-monogrid` to add `ne120pg2_ne120pg2` resolution, I ran 1 complete year with `F2010-SCREAMv1`.  Where above it was stopping after about 5.5 months with `ne120pg2_r0125_oRRS18to6v3`.

I also noticed a large increase in performance. On pm-gpu, using 32 nodes, the monogrid option was more than 2x faster (6.3 sypd vs 3 sypd). The increase was in the CPL and LND components.

```
/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/badd120/f120.F2010-SCREAMv1.ne120pg2_r0125_oRRS18to6v3.badd120.gnugpu.1m.n032a4xX

/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/badd120/f120.F2010-SCREAMv1.ne120pg2_ne120pg2.badd120.gnugpu.1m.n032a4xX
```
